controllers/etcdcluster_controller.go

```diff
@@ -62,22 +69,44 @@ func headlessServiceForCluster(cluster *etcdv1alpha1.EtcdCluster) *v1.Service {
 	}
 }
 
-// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters,verbs=get;list;watch
-// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters/status,verbs=get;update;patch
-// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create
-// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers,verbs=get;list;watch;create
+// updateStatus updates the EtcdCluster resource's status to be the current value of the cluster.
+func (r *EtcdClusterReconciler) updateStatus(
+	ctx context.Context,
+	name types.NamespacedName,
+	cluster *etcdv1alpha1.EtcdCluster,
+	members []etcdclient.Member,
+	reconcilerEvent event.EtcdClusterReconcilerEvent) error {
 
-func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-	defer cancel()
+	log := r.Log.WithValues("cluster", name)
 
-	log := r.Log.WithValues("cluster", req.NamespacedName)
+	if members != nil {
```

Comment:
are we confident that we're correctly handling nil vs empty here? looks like this slice originally comes from the etcd client, does that have the same semantics we assume?

Reply:
It's a solid point. I did consider a few alternatives:

1. Pass through the `error` from the member list as another parameter.
2. Bundle the `error` and `[]etcdclient.Member` into one struct and pass that.
3. Move the etcd client stuff into some struct with a `ListMembers` function. I could then impelement caching inside that struct and return the same members and error every time.

One seems messy as hell, two seems to go against the grain of what Go wants me to do (I'm really describing a 'maybe monad'), so if three seems reasonable I can do that?

Reply:
a fourth option could be to explicitly have this typed as a pointer, which we have complete control of, so we know when it's been set or not

Reply:
A pointer to a slice? Doesn't that end up as a pointer to a pointer? Or do you mean wrapping the `[]etcdclient.Member` slice in a struct and then passing a pointer to that struct?

Reply:
yep `*[]etcdclient.Member` - just makes it clearly obvious when we expect it to be nil vs empty

Reply:
I don't think you need to check for nil at all.
You you can get the length and loop over a nil slice: https://play.golang.org/p/cuzyctf-sYw

And we always want to set the `.Status.Members` field, even if its empty.

Reply:
I still don't get this.
If there's a network blip, the etcd client connection  / request will fail and we should exit the reconciliation and try again, right?
But happy go with it and  see how it works out .